File: trinet-sdk/src/main/kotlin/com/panoculon/trinet/sdk/recording/VtsFileWriter.kt
package com.panoculon.trinet.sdk.recording

import com.panoculon.trinet.sdk.io.BinaryWriter
import com.panoculon.trinet.sdk.model.VtsEntry
import java.io.BufferedOutputStream
import java.io.Closeable
import java.io.File
import java.io.FileOutputStream

/**
 * Writes a TRIVTS01 v2 sidecar file. Layout matches `the reference format spec:74-81`:
 *   header(32): magic[8] "TRIVTS01", version=2, frame_rate_milli, reserved[16].
 *   entries(24 each): frame_number(u32), sof_ts_ns(u64), venc_seq(u32), venc_pts_us(u64).
 */
class VtsFileWriter(
    private val file: File,
    fps: Float,
) : Closeable {

    companion object {
        val MAGIC = "TRIVTS01".toByteArray(Charsets.US_ASCII)
        const val VERSION = 2
        const val HEADER_SIZE = 32
    }

    private val frameRateMilli: Int = (fps * 1000f).toInt()
    private val out = BufferedOutputStream(FileOutputStream(file))
    var entryCount: Long = 0L
        private set

    init {
        writeHeader()
    }

    private fun writeHeader() {
        val w = BinaryWriter(HEADER_SIZE)
        w.bytes(MAGIC)
        w.u32(VERSION.toLong())
        w.u32(frameRateMilli.toLong())
        w.zeros(HEADER_SIZE - w.position())
        out.write(w.toByteArray())
    }

    @Synchronized
    fun append(entry: VtsEntry) {
        val w = BinaryWriter(VtsEntry.SIZE_BYTES)
        entry.writeTo(w)
        out.write(w.toByteArray())
        entryCount++
    }

    @Synchronized
    fun flush() { out.flush() }

    @Synchronized
    override fun close() {
        out.flush()
        out.close()
    }
}

File: ios/Sources/TrinetSDK/Recording/VtsFileWriter.swift
// VtsFileWriter.swift — writes a TRIVTS01 v2 sidecar of per-frame timestamps.
//
// TRIVTS01 v2 format (shared with the Android SDK and Linux toolchain — see
// ios/docs/PROTOCOLS.md and ios/README.md):
//   Header (32 bytes):
//     magic[8] = "TRIVTS01"
//     version  = uint32 (2)
//     frame_rate_milli = uint32 (fps * 1000)
//     reserved[16]
//   Per-frame entry (24 bytes):
//     frame_number, sof_timestamp_ns, venc_seq, venc_pts_us

import Foundation

public final class VtsFileWriter {
    public let url: URL
    private let handle: FileHandle
    public private(set) var entryCount: Int = 0

    public init(url: URL, fps: Double) throws {
        self.url = url
        if FileManager.default.fileExists(atPath: url.path) {
            try FileManager.default.removeItem(at: url)
        }
        FileManager.default.createFile(atPath: url.path, contents: nil, attributes: nil)
        self.handle = try FileHandle(forWritingTo: url)
        try handle.write(contentsOf: VtsFileWriter.serialize(fps: fps))
    }

    public func append(_ entry: VtsEntry) throws {
        var d = Data(); d.reserveCapacity(VtsEntry.binarySize)
        entry.encode(into: &d)
        try handle.write(contentsOf: d)
        entryCount += 1
    }

    public func close() throws {
        try handle.synchronize()
        try handle.close()
    }

    static func serialize(fps: Double) -> Data {
        var d = Data(); d.reserveCapacity(32)
        d.append("TRIVTS01".data(using: .ascii)!)         // 8 bytes
        var ver: UInt32 = 2; withUnsafeBytes(of: &ver) { d.append(contentsOf: $0) }
        var fpsMilli: UInt32 = UInt32(round(fps * 1000))
        withUnsafeBytes(of: &fpsMilli) { d.append(contentsOf: $0) }
        d.append(Data(count: 16))                          // reserved
        precondition(d.count == 32, "TRIVTS01 header must be exactly 32 bytes")
        return d
    }
}
